### PR TITLE
Ycheck that all methods have method type

### LINKED ISSUE #1046: Ycheck that all methods have either MethodicType or PolyType(MethodicType)

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #1046 . Please review @DarkDimius .

### CODE REVIEW

**Comment by odersky** [comment]
Why does this fail?

**Comment by liufengyun** [comment]
Some tests failed related to the tree checking in this commit. I'll investigate. Most likely I missed some valid types [here](https://github.com/lampepfl/dotty/pull/1205/files#diff-42c854f47732e7d455a7821137f52e7cR241).

File: src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 1]
```diff
@@ -237,6 +237,14 @@ class TreeChecker extends Phase with SymTransformer {
       tree
     }
 
+    /** Check that all methods have either MethodicType or PolyType(MethodicType) */
+    def isMethodType(pt: Type)(implicit ctx: Context): Boolean = pt match {
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: Failing tests suggest that you may be missing AnnotatedTypes here.

Comment (liufengyun) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: Thanks @DarkDimius , `AnnotatedType` is indeed missing here. But there's something more related to default parameters.

Following generated function for default parameters

```
    def copy$default$1[a]: a @uncheckedVariance =
      Some.this.x: a @uncheckedVariance
```

gets following type:

```
PolyType(List(a), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing), TypeRef(ThisType(TypeRef(NoPrefix,scala)),Any))), AnnotatedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Some)),Some$$a),ConcreteAnnotation(Apply(Select(New(TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,unchecked)),uncheckedVariance)]),<init>),List()))))
```

Is it valid for methods of default parameters to take a non-method type?

**Comment by liufengyun** [comment]
retest this please

**Comment by DarkDimius** [comment]
LGTM

**Pull request merged**